magicblock-committor-service/src/commit_strategy.rs

```diff
@@ -0,0 +1,633 @@
+pub struct SplitChangesets {
+    /// This changeset can be committed in one processing step, passing account data as args
+    pub args_changeset: Changeset,
+    /// This changeset can be committed in one processing step, passing account data as args
+    /// and the finalize instruction fits into the same transaction
+    pub args_including_finalize_changeset: Changeset,
+    /// This changeset can be committed in one processing step, passing account data as args
+    /// but needs to use lookup tables for the accounts
+    pub args_with_lookup_changeset: Changeset,
+    /// This changeset can be committed in one processing step, passing account data as args
+    /// and the finalize instruction fits into the same transaction.
+    /// It needs to use lookup tables for the accounts.
+    pub args_including_finalize_with_lookup_changeset: Changeset,
+    /// This changeset needs to be committed in two steps:
+    /// 1. Prepare the buffer account
+    /// 2. Process the buffer account
+    pub from_buffer_changeset: Changeset,
+    /// This changeset needs to be committed in three steps:
+    /// 1. Prepare the buffer account
+    /// 2. Prepare lookup table
+    /// 3. Process the buffer account
+    pub from_buffer_with_lookup_changeset: Changeset,
+}
+
+pub fn split_changesets_by_commit_strategy(
+    changeset: Changeset,
+    finalize: bool,
+) -> CommittorServiceResult<SplitChangesets> {
+    fn add_to_changeset(
+        changeset: &mut Changeset,
+        accounts_to_undelegate: &HashSet<Pubkey>,
+        bundle: ChangedBundle,
+    ) {
+        for (pubkey, acc) in bundle {
+            changeset.add(pubkey, acc);
+            if accounts_to_undelegate.contains(&pubkey) {
+                changeset.accounts_to_undelegate.insert(pubkey);
+            }
+        }
+    }
+
+    let mut args_changeset = Changeset {
+        slot: changeset.slot,
+        ..Default::default()
+    };
+    let mut args_including_finalize_changeset = Changeset {
+        slot: changeset.slot,
+        ..Default::default()
+    };
+    let mut args_with_lookup_changeset = Changeset {
+        slot: changeset.slot,
+        ..Default::default()
+    };
+    let mut args_including_finalize_with_lookup_changeset = Changeset {
+        slot: changeset.slot,
+        ..Default::default()
+    };
+    let mut from_buffer_changeset = Changeset {
+        slot: changeset.slot,
+        ..Default::default()
+    };
+    let mut from_buffer_with_lookup_changeset = Changeset {
+        slot: changeset.slot,
+        ..Default::default()
+    };
+
+    let accounts_to_undelegate = changeset.accounts_to_undelegate.clone();
+    let changeset_bundles = changeset.into_small_changeset_bundles();
+    for bundle in changeset_bundles.bundles.into_iter() {
+        let commit_strategy =
+            CommitBundleStrategy::try_from((bundle, finalize))?;
+        match commit_strategy {
+            CommitBundleStrategy::Args(bundle) => {
+                add_to_changeset(
```

Comment:
is it safe to append on new iteration to same `args_changeset`?

Say we have 2 bundles:
Iteration 1 - we get CommitBundleStrategy::Args and push to args_changeset.
Iteration 2 - we get CommitBundleStrategy::Args and push to args_changeset.

Now the issue it that if bundle in Iteration 1 and 2 separately fit into args, doesn't mean they both will.

Looking into `commit_changeset_using_args` it doesn't handle them on bundle levels but tries create a single tx with data from 2 Iterations